### ISSUE TITLE
Multipart parsing error - Final boundary missing

### ISSUE THREAD

**Issue by devnull-hub-lab** [SpiderLabs/ModSecurity#2733]
Hi guys,

I have a problem here.
I'm using MantisBt Bugtracker system. When modsecurity is activem and I submitting a form, a receive error: "Bad Request".

In audit logs, in resuming I got:
ModSecurity: Access denied with code 400 (phase 2). Matched "Operator `Eq' with parameter `0' against variable `REQBODY_ERROR' (Value: `1' ) [msg "Failed to parse request body."] [data "Multipart parsing error: Multipart: Final boundary missing."]

In debug (level 9), I got:
....
[1646250438] [/abc71/bugtracker/bug_report.php] [9] Multipart: Added data to variable: 
[1646250438] [/abc71/bugtracker/bug_report.php] [1] Multipart: Final boundary missing.
[1646250438] [/abc71/bugtracker/bug_report.php] [4] Multipart: Cleanup started (keep files set to Not set)
[1646250438] [/abc71/bugtracker/bug_report.php] [9] This phase consists of 338 rule(s).
[1646250438] [/abc71/bugtracker/bug_report.php] [4] (Rule: 200002) Executing operator "Eq" with param "0" against REQBODY_ERROR.
[1646250438] [/abc71/bugtracker/bug_report.php] [9] Target value: "1" (Variable: REQBODY_ERROR)
[1646250438] [/abc71/bugtracker/bug_report.php] [9] Matched vars updated.
[1646250438] [/abc71/bugtracker/bug_report.php] [4] Rule returned 1.

I could not identify the reason for the error, that is, what is missing?

I contacted the MantisBt Bugtracker team, and they didn't identify any bugs, pointing to modsecurity possible issue.
And indeed, with modsecurity disabled, submit works fine.

I attached the complete audit log  and debug log from this request.
[modsec_debug.log](https://github.com/SpiderLabs/ModSecurity/files/8585648/modsec_debug.log)
[modsec_audit.log](https://github.com/SpiderLabs/ModSecurity/files/8585647/modsec_audit.log)

**Server**
 - ModSecurity version 3.0.6 (with modsecurity3-apache-0.0.9 connector)
 - WebServer: Apache 2.4.53
 - OS: FreeBSD 13.0-RELEASE-p10

**Rule Set**
 - Running public rule set: https://github.com/coreruleset/coreruleset/tree/v3.4/dev
 - Version number: [2022-03-15]

I saw that there is an issue related to limitrequestbody: #2670
I don't know if it has anything to do with this issue.

I would be happy if someone can help me, and I'm available for any test needed.

Thanks

**Comment by martinhsv** [comment]
Hello @devnull-hub-lab ,

I'll caution you right away that using ModSecurity v3 with the apache connector is not recommended. While much of the code works correctly, not all of it does, and it's definitely not considered to be production-ready. For use with Apache HTTP Server, ModSecurity v2.9.x is recommended.

A few additional notes:
- you have found #2670 . If you are using the tagged version of v3.0.6 (from last November) that functionality is not present. If you are using current v3/master, then it is present, but still unlikely to be related to your issue
- request body size limits, can, in principal, trigger that sort of effect but, you have to be using ```SecRequestBodyLimitAction ProcessPartial``` for that to be a likely effect. Moreover, based on your debug log, there does not appear to be any such issue ("[1646250438] [/abc71/bugtracker/bug_report.php] [9] Appending request body: 2166 bytes. Limit set to: 13107200.000000"  )

With the information available, it look like the two most likely explanations are:
- the final boundary was indeed missing from the request body
- the final boundary only appeared (to ModSecurity) to be missing due to some bug related to the apache connector

**Comment by devnull-hub-lab** [comment]
Hi @martinhsv,

Thanks for you reply and attention!

I've been using modsecurity 3.x for about 1,5 year, and I've never had any problems, with the exception of this one. And I don't really know if it would have anything to do with the apache connector. But thanks for advice.

At first I installed modsecurity 2.x, but I had unpleasant surprises, with modsec_audit.log generating strange and aleatory characters. This way I ended up installing modsecurity 3.x, and solved this problem.

Booth (modsec 2.x and 3.x) are still maintained and are full compatible with [coreruleset](https://github.com/coreruleset/coreruleset) ?

After your considerations:
1 - If final boundary was indeed missing from the request body, then the same error will happen with modsecurity 2.x.
2 - If exists some bug in apache connector, installing modsecurity 2.x will solve the problem.

Both cases seem to point to modsecurity 2.x as a solution and diagnosis.

But before installing version 2.x, since I have modsecurity 3.0.6, I will apply the changes from #2670 and #2731, maybe it has something to do with it, it's worth a try.

That way, I still keep you updated, whether the results are positive or negative.

Thanks again for your attention!
